software/bootcamp-code/main.py

```diff
@@ -0,0 +1,137 @@
+import image_processor
+import camera
+import motion
+import cv2
+import time
+
+# function to move wheel/wheels for given time
+def move_wheel_s(omnim_robot_instance, start_time, end_time, time_passed, speed1=0, speed2=0, speed3=0):
+    if time_passed<start_time:
+        return
+    elif time_passed>end_time:
+        return
+    else:
+        omnim_robot_instance.send_commands(speed1, speed2, speed3, 0)
+
+def main_loop():
+    debug = True
+    
+    #motion_sim = motion.TurtleRobot()
+    #motion_sim2 = motion.TurtleOmniRobot()
+
+    omni_motion = motion.OmniMotionRobot()
+    
+    #camera instance for normal web cameras
+    #cam = camera.OpenCVCamera(id = 2)
+    #camera instance for realsense cameras
+    cam = camera.RealsenseCamera(exposure = 100)
+    processor = image_processor.ImageProcessor(cam, debug=debug)
+
+    processor.start() # ilma kaamerata hangub siin
+
+    #motion_sim.open()
+    #motion_sim2.open()
+
+    # open the serial connection
+    omni_motion.open()
+
+    start = time.time()
+    fps = 0
+    frame = 0
+    frame_cnt = 0
+    try:
+        start_time = time.time()
+        state = 0
+        while True:
+            # STATE MACHINE: 3 states - FIND_BALL = 0, DRIVE_TO_BALL = 1, ALREADY_HAVE_A_BALL = 2
+
+            # has argument aligned_depth that enables depth frame to color frame alignment. Costs performance
+            processedData = processor.process_frame(aligned_depth=False)
+
+            # TODO This is where you add the driving behaviour of your robot. It should be able to filter out
+            # objects of interest and calculate the required motion for reaching the objects
+            print()
+
+            if state>1:
```

Comment:
I suggest making states control state transitions, then all logic is consolidated in one place. The robot can always start with `find ball` and if it instantly finds a ball, the find ball state sets the new state to `drive to ball`. If that state has no balls to drive to, it will transition back to finding a ball.